### ISSUE TITLE
Our FECache code doesn't work with subdivision elements

### ISSUE THREAD

**Issue by drwells** [IBAMR/IBAMR#675]
@xxjaehoxx This is the issue we talked about today :)

This is not so much a problem with FECache, but the way we use it:
```cpp
FEBase& X_fe = X_fe_cache[key];
QBase& qrule = d_quadrature_cache[key];

// See the note in interpWeighted to explain why we override
// libMesh's reinit logic here
if (used_quadratures.find(key) == used_quadratures.end())
{
    X_fe.get_phi();
    X_fe.reinit(elem);
    used_quadratures.insert(key);
}
const std::vector<std::vector<double> >& X_phi = X_fe.get_phi();
```
i.e., we get the (possibly uninitialized) object from the cache and then only initialize it if we have a new quadrature rule. This won't work with subdivision elements since they have, in general, a different number of degrees of freedom per element (i.e., we can keep the same quadrature rule but the `phi` array will still be different).

I think that the correct way to fix this is to instruct FECache to also initialize the FE objects it stores appropriately: e.g., if we have subdivision elements then we always need to recompute shape functions, but if we have Lagrange elements we only need to do it once.

**Comment by boyceg** [comment]
Can we close this issue?

**Comment by xxjaehoxx** [comment]
Sounds good to me!

**Comment by drwells** [comment]
The fix in #679 should be sufficient; we can always open this again if we want to do more work with subdivision elements.
